src/traversal.ts

```diff
@@ -0,0 +1,42 @@
+import { $URL } from './url'
+import { withTrailingSlash } from './utils'
+
+interface TraversePathTreeRuntime {
+  /**
+   * Whether we need to append a trailing slash to each path node.
+   */
+  trailingSlash?: boolean
+  /**
+   * The URL nodes found when traversing
+   */
+  nodes: string[]
+}
+
+/**
+ * Steps through a URL removing the last segment each time, giving you a tree of urls leading up to the provided url.
+ *
+ * For example, the traversal for the URL /blog/article/my-long-article-title will be:
+ * - /blog/article/my-long-article-title
+ * - /blog/article
+ * - /blog
+ * - /
+ */
+export const traversePathTree = (url: string, runtime: TraversePathTreeRuntime = { nodes: [] }) => {
```

Comment:
Why not directly return (immutable) results? (We might also use a generator function or array iterator to allow traverse one by one)